[SPARK-26482][CORE] Use ConfigEntry for hardcoded configs for ui categories

HeartSaVioR opened the pull request:
## What changes were proposed in this pull request?

The PR makes hardcoded configs below to use `ConfigEntry`.

* spark.ui
* spark.ssl
* spark.authenticate
* spark.master.rest
* spark.master.ui
* spark.metrics
* spark.admin
* spark.modify.acl

This patch doesn't change configs which are not relevant to SparkConf (e.g. system properties).

## How was this patch tested?

Existing tests.

HeartSaVioR committed:
[SPARK-26482][CORE] Use ConfigEntry for hardcoded configs for ui categories

* Rename conflicted configuration field names
* Move out UI related configurations to separate object
* Address review comments from @vanzin

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #100619 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100619/testReport)** for PR 23423 at commit [`016ebe0`](https://github.com/apache/spark/commit/016ebe0a2987a6e534090cb0aef9e4dd8b630caa).

SparkQA commented:
**[Test build #100619 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100619/testReport)** for PR 23423 at commit [`016ebe0`](https://github.com/apache/spark/commit/016ebe0a2987a6e534090cb0aef9e4dd8b630caa).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `    .doc(\"Comma separated list of filter class names to apply to the Spark Web UI.\")`
  * `    .doc(\"Comma-separated list of class names for \"plugins\" implementing \" +`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100619/
Test FAILed.

SparkQA commented:
**[Test build #100621 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100621/testReport)** for PR 23423 at commit [`1c9ee36`](https://github.com/apache/spark/commit/1c9ee369b80a746fb42f2a21c781a2eec776970e).

SparkQA commented:
**[Test build #100622 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100622/testReport)** for PR 23423 at commit [`5973c59`](https://github.com/apache/spark/commit/5973c59936b7fe266f9d04c45f230a47b8730542).
 * This patch **fails to build**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100622/
Test FAILed.

SparkQA commented:
**[Test build #100624 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100624/testReport)** for PR 23423 at commit [`8d8721d`](https://github.com/apache/spark/commit/8d8721dc9fd2ca59333faf8e6ebb4d4812fa6cc1).

SparkQA commented:
**[Test build #100624 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100624/testReport)** for PR 23423 at commit [`8d8721d`](https://github.com/apache/spark/commit/8d8721dc9fd2ca59333faf8e6ebb4d4812fa6cc1).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100624/
Test FAILed.

SparkQA commented:
**[Test build #100626 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100626/testReport)** for PR 23423 at commit [`6dc46aa`](https://github.com/apache/spark/commit/6dc46aa488f4846797ccd0a7b05fbe2ecd560e88).

SparkQA commented:
**[Test build #100626 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100626/testReport)** for PR 23423 at commit [`6dc46aa`](https://github.com/apache/spark/commit/6dc46aa488f4846797ccd0a7b05fbe2ecd560e88).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100626/
Test FAILed.

SparkQA commented:
**[Test build #100627 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100627/testReport)** for PR 23423 at commit [`15a8955`](https://github.com/apache/spark/commit/15a8955364f7d47ceddc9086320e68cdd3eb1a99).

SparkQA commented:
**[Test build #100627 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100627/testReport)** for PR 23423 at commit [`15a8955`](https://github.com/apache/spark/commit/15a8955364f7d47ceddc9086320e68cdd3eb1a99).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100627/
Test FAILed.

HeartSaVioR commented:
Strange error occurred after rebasing. Changed to WIP. Will fix the error and change the title back.

SparkQA commented:
**[Test build #100638 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100638/testReport)** for PR 23423 at commit [`caa79d1`](https://github.com/apache/spark/commit/caa79d11eef5aceafd21c5e9b534fe8aab3a1540).

SparkQA commented:
**[Test build #100638 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100638/testReport)** for PR 23423 at commit [`caa79d1`](https://github.com/apache/spark/commit/caa79d11eef5aceafd21c5e9b534fe8aab3a1540).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100638/
Test FAILed.

SparkQA commented:
**[Test build #100646 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100646/testReport)** for PR 23423 at commit [`2569a75`](https://github.com/apache/spark/commit/2569a75e0ba230a22b25c59dc90de8eaee62a5cd).

SparkQA commented:
**[Test build #100646 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100646/testReport)** for PR 23423 at commit [`2569a75`](https://github.com/apache/spark/commit/2569a75e0ba230a22b25c59dc90de8eaee62a5cd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100646/
Test PASSed.

vanzin commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -597,6 +611,53 @@ package object config {
       .toSequence
       .createWithDefault(Nil)
 
+  private[spark] val UI_ENABLED = ConfigBuilder("spark.ui.enabled")
```
Could you put UI-related options in a new file (e.g. `UI.scala`)?

I was going to suggest `WebUI.scala` but that already exists (albeit in a different place).

HeartSaVioR replied:
Thanks for the nice suggestion. Will address.

SparkQA commented:
**[Test build #100667 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100667/testReport)** for PR 23423 at commit [`31ba1d6`](https://github.com/apache/spark/commit/31ba1d6e9e772d10ea2e2cca8e57c013b2ddbf35).

SparkQA commented:
**[Test build #100667 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100667/testReport)** for PR 23423 at commit [`31ba1d6`](https://github.com/apache/spark/commit/31ba1d6e9e772d10ea2e2cca8e57c013b2ddbf35).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `    .doc(\"Comma separated list of filter class names to apply to the Spark Web UI.\")`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100667/
Test FAILed.

SparkQA commented:
**[Test build #100671 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100671/testReport)** for PR 23423 at commit [`38dc7be`](https://github.com/apache/spark/commit/38dc7be0a150ce4f5e7ecaf3beb1e23036748cc1).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100671/
Test FAILed.

SparkQA commented:
**[Test build #100678 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100678/testReport)** for PR 23423 at commit [`38dc7be`](https://github.com/apache/spark/commit/38dc7be0a150ce4f5e7ecaf3beb1e23036748cc1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100678/
Test PASSed.

vanzin commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -724,13 +709,44 @@ package object config {
+    .intConf
+    .createWithDefault(8080)
+
+  private[spark] val ADMIN_ACLS = ConfigBuilder("spark.admin.acls")
```
In spite of the name, all of these ACL-related options are UI options.

HeartSaVioR replied:
I thought these options are common thing between UI and History, but looks like History has its own ACL configurations. Will move to UI.

vanzin commented on core/src/main/scala/org/apache/spark/internal/config/UI.scala:
```diff
@@ -0,0 +1,120 @@
+
+private[spark] object UI {
+
+  val UI_SHOW_CONSOLE_PROGRESS = ConfigBuilder("spark.ui.showConsoleProgress")
```
I'm leaning a bit on the side of removing the `UI_` prefix here, and always referencing these as `UI.foo` in other code.

But don't really feel strongly either way.

HeartSaVioR replied:
Sounds nice. I'd like to ensure "how to use configurations" be consistent though: breaking down config package object into multiple (like what we do with UI, History, Kafka, etc) and referring configuration as `<object name>.<configuration name>` all the cases. 

Would it make sense to file issues for cleanup (more groups, change the way to refer configurations) as part of [SPARK-26442](https://issues.apache.org/jira/browse/SPARK-26442)?

vanzin replied:
Someone filed SPARK-22194 a while ago that is in that same spirit.

As for consistency, it may be better to be even more explicit (e.g. `config.Blah.CONSTANT`) but that is pretty verbose. Some parts of the code already do that.

vanzin commented on core/src/main/scala/org/apache/spark/internal/config/UI.scala:
```diff
@@ -0,0 +1,120 @@
+    .stringConf
+    .createWithDefaultString("")
+
+  val UI_ACLS_ENABLE = ConfigBuilder("spark.ui.acls.enable")
```
Since you're replacing them with the constant anyway, I'd rather have the tests use the new constant.

Then this backwards compatibility could be added to the deprecated stuff in the `SparkConf` object. But really, given the comment in `SecurityManager`, maybe we could even drop it altogether.

HeartSaVioR replied:
+1 to drop it, since it's for backward compatibility with 1.x and target version of this patch is 3.0 (new major version). Will drop it.

vanzin commented on core/src/main/scala/org/apache/spark/ui/SparkUI.scala:
```diff
@@ -151,12 +152,11 @@ private[spark] abstract class SparkUITab(parent: SparkUI, prefix: String)
 
   def getUIPort(conf: SparkConf): Int = {
-    conf.getInt("spark.ui.port", SparkUI.DEFAULT_PORT)
+    conf.get(UI_PORT)
```
Can `SparkUI.DEFAULT_PORT` go away?

HeartSaVioR replied:
Yes it is only used for default value on configuration for `UI_PORT` at least in Spark codebase, so I guess it can be removed once `UI_PORT` has default value.

vanzin commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -724,13 +709,44 @@ package object config {
+    .toSequence
+    .createWithDefault(Nil)
+
+  private[spark] val MASTER_REST_SERVER_ENABLED = ConfigBuilder("spark.master.rest.enabled")
```
These could be in a `Standalone` object, but since they're few, probably ok to have them here for now.

HeartSaVioR replied:
I'm OK in either way. I might miss some more configurations on standalone mode so we may have chance to move them in future.

SparkQA commented:
**[Test build #100721 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100721/testReport)** for PR 23423 at commit [`6bb772d`](https://github.com/apache/spark/commit/6bb772d0e4b0897dfc564574b532f82f27df743a).

SparkQA commented:
**[Test build #100721 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100721/testReport)** for PR 23423 at commit [`6bb772d`](https://github.com/apache/spark/commit/6bb772d0e4b0897dfc564574b532f82f27df743a).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100721/
Test FAILed.

HeartSaVioR commented:
retest this, please

SparkQA commented:
**[Test build #100725 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100725/testReport)** for PR 23423 at commit [`6bb772d`](https://github.com/apache/spark/commit/6bb772d0e4b0897dfc564574b532f82f27df743a).

SparkQA commented:
**[Test build #100725 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100725/testReport)** for PR 23423 at commit [`6bb772d`](https://github.com/apache/spark/commit/6bb772d0e4b0897dfc564574b532f82f27df743a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100725/
Test PASSed.

kiszk commented on core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala:
```diff
@@ -304,11 +305,10 @@ object HistoryServer extends Logging {
+    if (config.get(ACLS_ENABLE)) {
+      logInfo(s"${ACLS_ENABLE.key} is configured, " +
+        s"clearing it and only using ${History.HISTORY_SERVER_UI_ACLS_ENABLE.key}")
+      config.set(ACLS_ENABLE.key, "false")
```
How about `config.set(ACLS_ENABLE, false)`?

HeartSaVioR replied:
Nice finding! Will address.

HeartSaVioR committed:
Address @kiszk review comment

HeartSaVioR commented:
@vanzin @kiszk 
Addressed review comments. Seems like PRs for subtask of SPARK-26442 can make each other being conflicted, so it might be ideal if we could take review and merge a bit faster, and address some kinds of TODOs (like cleaning up) after all done.

SparkQA commented:
**[Test build #100855 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100855/testReport)** for PR 23423 at commit [`6edcc6f`](https://github.com/apache/spark/commit/6edcc6fd93fd97f992d34c0b648f7c6f47d64182).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100855/
Test PASSed.

vanzin commented on core/src/main/scala/org/apache/spark/internal/config/UI.scala:
```diff
@@ -0,0 +1,138 @@
+    .intConf
+    .createWithDefault(4040)
+
+  val UI_FILTERS = ConfigBuilder("spark.ui.filters")
```
This could use `.toSequence` too and clean up call sites a little bit.

HeartSaVioR replied:
Thanks for the suggestion. Looks like others which are used along with `stringToSet` can be applied. I'll see which configurations can be applied and make changes.

gaborgsomogyi commented:
Just wanted to check, could you merge the master?

HeartSaVioR commented:
Just rebased (with addressing config changes to the new code), still need to reflect @vanzin's latest comment.

SparkQA commented:
**[Test build #100987 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100987/testReport)** for PR 23423 at commit [`a3f06e8`](https://github.com/apache/spark/commit/a3f06e841a666be739a9782a3674d233633e0c86).

HeartSaVioR committed:
Address review comment from @vanzin

SparkQA commented:
**[Test build #100990 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100990/testReport)** for PR 23423 at commit [`b3526ff`](https://github.com/apache/spark/commit/b3526ffb63106a98d07e12f96a2ca1b525d6de62).

HeartSaVioR committed:
Also apply `.toSequence` to configurations in UI and History

SparkQA commented:
**[Test build #100994 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100994/testReport)** for PR 23423 at commit [`8d61242`](https://github.com/apache/spark/commit/8d61242aa85661592fc530a7cd5a8956f6f75bb8).

SparkQA commented:
**[Test build #100987 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100987/testReport)** for PR 23423 at commit [`a3f06e8`](https://github.com/apache/spark/commit/a3f06e841a666be739a9782a3674d233633e0c86).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100987/
Test PASSed.

SparkQA commented:
**[Test build #100994 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100994/testReport)** for PR 23423 at commit [`8d61242`](https://github.com/apache/spark/commit/8d61242aa85661592fc530a7cd5a8956f6f75bb8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100994/
Test PASSed.

HeartSaVioR commented:
@kiszk @vanzin 
I guess I addressed review comments and the build is green. 
Could you please take a look again? Thanks in advance!

SparkQA commented:
**[Test build #101022 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101022/testReport)** for PR 23423 at commit [`f5a5772`](https://github.com/apache/spark/commit/f5a5772c109869cf9480c06052a1a9c43f5bc575).

SparkQA commented:
**[Test build #101022 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101022/testReport)** for PR 23423 at commit [`f5a5772`](https://github.com/apache/spark/commit/f5a5772c109869cf9480c06052a1a9c43f5bc575).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101022/
Test FAILed.

HeartSaVioR commented:
retest this, please

SparkQA commented:
**[Test build #101036 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101036/testReport)** for PR 23423 at commit [`f5a5772`](https://github.com/apache/spark/commit/f5a5772c109869cf9480c06052a1a9c43f5bc575).

vanzin commented:
Looks good pending tests.

SparkQA commented:
**[Test build #101036 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101036/testReport)** for PR 23423 at commit [`f5a5772`](https://github.com/apache/spark/commit/f5a5772c109869cf9480c06052a1a9c43f5bc575).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101036/
Test PASSed.

vanzin commented:
Merging to master.

dongjoon-hyun commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesTestComponents.scala:
```diff
@@ -67,8 +68,8 @@ private[spark] class KubernetesTestComponents(defaultClient: DefaultKubernetesCl
-      .set("spark.ui.enabled", "true")
-      .set(IS_TESTING.key, "false")
+      .set(IS_TESTING, false)
+      .set(UI_ENABLED, true)
```
Hi, all. This breaks the K8S integration test.
I'll make a followup PR.
